src/widget-core/tsx.ts

```diff
@@ -54,3 +64,51 @@ export function tsx(tag: any, properties = {}, ...children: any[]): DNode {
 		return w(tag, properties, children);
 	}
 }
+
+function createFactory(callback: any, middlewares: any): any {
+	const factory = (properties: any, children?: any) => {
+		if (properties) {
+			const result = w(callback, properties, children);
+			callback.isWidget = true;
+			callback.middlewares = middlewares;
+			return result;
+		}
+		return {
+			middlewares,
+			callback
+		};
+	};
+	factory.isFactory = true;
+	return factory;
+}
+
+export function create<T extends MiddlewareMap<any>, MiddlewareProps = ReturnType<T[keyof T]>['properties']>(
+	middlewares: T = {} as T
+) {
+	function properties<Props extends {}>() {
```

Comment:
Would this read better as something similar to `withProperties()`? Thinking about it from a perspective of making the create API more fluent, e.g.:

```tsx
create().withProperties<{]>();
```

Reply:
I was trying to reduce the number of characters but when with `properties`, I don't actually mind the `with` prefix tho for it's fluency. @matt-gadd any thoughts?

Reply:
There's another subtle comprehension hiccup here - this method is describing the property interface as part of the factory creation, not widget creation nor rendering invocation.

Reading it as it currently stands, a user could potentially confuse it with passing actual property values as part of a 'create a widget` call; something that `withProperties` or `hasProperties` doesn't help with to resolve.

What about something like `withAPI()`?

Reply:
My issue with that is that in the future, when typescript generics catch up we’d like to explore the possibility of using the function to define default properties which will leave the external property API with optional properties but enable referencing them from within widget knowing that although the external API is optional the internal API is not. For the moment typescript would prevent a user passing anything to the function as there are no arguments current.

Additionally the API of a widget is not just properties, it’s also children which again in the future is something that we’d consider exposing for the user to define perhaps with a further `.children` API.

Reply:
Makes sense.